---
title: "rcycle: PCA of Transcriptome Time Series"
author: "Rainer Machn&eacute;"
date: "`r format(Sys.time(), '%d %m %Y')`"
output:
    bookdown::html_document2:
        base_format: rmarkdown::html_vignette
        toc: true
        toc_depth: 2
        fig_caption: true
bibliography: /home/raim/ref/tata.bib
vignette: >
  %\VignetteIndexEntry{Piecewise Linear Segmentation by Dynamic Programming}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<!-- compile with
rmarkdown::render("vignettes/rcycle.Rmd", output_format = "html_document")
-->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, global.par=TRUE,
                      fig.path = ".", fig.pos = 'h',
                      fig.height = 2.7, fig.width = 4, fig.align = "center")
knitr::opts_knit$set(global.par = TRUE)
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25, cex.main=.75)
require(rcycle)
require(segmenTools)
source("~/programs/rcycle/R/phases.R")
source("~/programs/rcycle/R/plot.R")
```





$$
\newcommand{\Ell}{\mathcal{L}}
\newcommand{\jump}{\mathcal{J}}
\newcommand{\Var}{\mathrm{Var}}
\newcommand{\Cov}{\mathrm{Cov}}
\newcommand{\lmax}{\ell_\text{max}}
\newcommand{\lmin}{\ell_\text{min}}
\newcommand{\def}{\stackrel{\mathrm{def}}{=}}
$$

TODO:

* FILL THIS WITH teaching/pca/pca.md, teaching/pseudophase/cohorts14.md
  and clusterTranscriptomes_understandPCA.R
* use R's svd notation for matrices, or @Alter2000.

# Singular Value Decomposition and Principal Component Analysis


```{r load, echo=FALSE, include=TRUE}
## run pca in various versions, using data by Li and Klevecz (2006)
script_path <- system.file("scripts", "li06_pca.R", package = "yourpackage")
script_path <- "/home/raim/programs/rcycle/inst/scripts/li06_pca.R"
if ( !exists("cphase") )
source(script_path)
```

```{r}
### GENE-CENTERED
## VARIABLES/COLUMNS are samples/cells, OBSERVATIONS/ROWS are genes
## row-normalized and UNTRANSPOSED
gdat <- pdat - apply(pdat, 1, mean)
gphase <- prcomp(gdat, scale.=TRUE)

### SAMPLE-CENTERED with genes as columns!
## VARIABLES/COLUMNS are genes, OBSERVATIONS/ROWS are samples/cells

## row-norm over cells
sdat <- t(pdat) # transpose BEFORE row-norm 
sdat <- sdat - apply(sdat, 1, mean)  # row-norm over cells!
sphase <- prcomp(sdat, scale.=TRUE)

## row-norm over genes: as in @Schwabe2000?
cdat <- pdat - apply(pdat, 1, mean)  # row-norm over genes
cdat <- t(cdat) # transpose AFTER row-norm 
cphase <- prcomp(cdat, scale.=FALSE)  # NOTE: -> colnorm!

## pseudophases for all approaches
gtheta <- atan2(gphase$rotation[,2], gphase$rotation[,1])
stheta <- atan2(sphase$x[,2], sphase$x[,1])
ctheta <- atan2(cphase$x[,2], cphase$x[,1])
```

# The `biplot`

* @Gabriel_1971
* (https://stats.stackexchange.com/questions/66926/what-are-the-four-axes-on-pca-biplot)
* (https://stats.stackexchange.com/questions/141085/positioning-the-arrows-on-a-pca-biplot)


```{r, echo=TRUE, fig.width=7, fig.height=3.5, fig.align='left', fig.cap='plotPC is equivalent to biplot'}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 2))

## TODO: test biplot-like scaling more exactly!
## biplot w/o scaling
plotPC(gphase, sarrows=FALSE, scores=TRUE, vlines=TRUE, vaxis=TRUE,
       scale=0, col=2, pch=19, cex=1)
par(new=TRUE)
biplot(gphase, scale=0, xlabs = rep("x", nrow(gphase$x)),
       ylabs = rep("-", ncol(gphase$x)), col=5, xlab=NA, ylab=NA,
       var.axes=FALSE) 
figlabel('unscaled', pos='bottomleft', font=2)

## biplot with scaling
plotPC(gphase, sarrows=FALSE, scores=TRUE, vlines=TRUE, vaxis=TRUE,
       scale=1, arcsinh=FALSE, col=2, pch=19, cex=1)
par(new=TRUE)
biplot(gphase, scale=1, xlabs = rep("x", nrow(gphase$x)),
       ylabs = rep("-", ncol(gphase$x)), col=5, xlab=NA, ylab=NA,
       var.axes=FALSE)
figlabel('true biplot', pos='bottomleft', font=2)
```

# Gene vs. Sample PCA


```{r, echo=TRUE, fig.width=9, fig.height=3, fig.align="left", fig.cap="PCA of a periodic transcriptome time series, either taken from the gene X sample (left) or the transposed sample X gene (middle and right) matrix; red: eigenvectors; black: PC scores. Lines connect samples in their temporal order to emphasize how PCA reconstructs the temporal order."}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))

## sample PCA: with samples as variables/columns - most unconventional
plotPC(gphase, sarrows=FALSE, scores=TRUE, vlines=TRUE, vaxis=TRUE,
       col=2, pch=19, cex=2) # eigenvectors/rotation matrix are samples
figlabel('sample PCA', pos='bottomleft', font=2)

## gene PCA: with genes as variables/columns
## but with row-norm over cells after transpose
plotPC(sphase, sarrows=FALSE, scol=1, spch=19, scex=2, col=2, vaxis=TRUE)
lines(sphase$x[,1], sphase$x[,2], col=1) # rotated data are samples
text(sphase$x[,1], sphase$x[,2], labels=1:nrow(sphase$x), col='white', cex=.8)
figlabel('gene PCA, cell-norm', pos='bottomleft', font=2)

## gene PCA: with genes as variables/columns - conventional
## with row-norm over genes prior to transpose
plotPC(cphase, sarrows=FALSE, scol=1, spch=19, scex=2, col=2, vaxis=TRUE)
lines(cphase$x[,1], cphase$x[,2], col=1) # rotated data are samples
text(cphase$x[,1], cphase$x[,2], labels=1:nrow(cphase$x), col='white', cex=.8)
figlabel('gene PCA, gene-norm', pos='bottomleft', font=2)
```

```{r, echo=TRUE}
## COMPARE PSEUDOPHASES
## pseudophases are very similar between approaches
gtheta <- atan2(gphase$rotation[,2], gphase$rotation[,1])
stheta <- atan2(sphase$x[,2], sphase$x[,1])
ctheta <- atan2(cphase$x[,2], cphase$x[,1])
```

```{r, echo=FALSE, fig.width=9, fig.height=3}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))
plot(time, gtheta, xlab='time/h', ylab=expression(theta~'sample PCA'))
plot(time, ctheta, xlab='time/h', ylab=expression(theta~'gene PCA, cell-norm'))
plot(time, stheta, xlab='time/h', ylab=expression(theta~'gene PCA, gene-norm'))
```

```{r, echo=FALSE, fig.width=9, fig.height=3}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))
plot(gtheta, stheta, xlab=expression(theta~'sample PCA'), ylab=expression(theta~'gene PCA, cell-norm'))
plot(gtheta, ctheta, xlab=expression(theta~'sample PCA'), ylab=expression(theta~'gene PCA, gene-norm'))
plot(stheta, ctheta, xlab=expression(theta~'gene PCA, cell-norm'), ylab=expression(theta~'gene PCA, gene-norm'))
```

TODO:

* same on cohort level!
* single cell data: are we using with or w/o rownorm?
* why does log(counts+1) appear to have no effect (tested in cohorts_cells.R)?


# Time Series Reconstruction: `pseudophase`.

eigengenes @Alter2000,

applied for single cell analysis: @Schwabe2020, @Macosko2015


# PCA vs. SVD

Same thing, but in practice people perform PCA to look at the rotated data
and view this as a dimensionality reduction. When using SVD people often
focus on the eigenvectors They are equivalent, and e.g. R's `prcomp` internally used the `svd` function (based on LAPACK routines‘DGESDD’ and ‘ZGESDD’

```{r, echo=TRUE, include=TRUE, fig.width=6, fig.height=3}

# Simulate centered gene expression data
set.seed(1)
n <- 500 # genes
m <- 12   # samples
X <- matrix(rnorm(n*m), nrow = n, ncol = m) # 500 genes X 5 samples

## TODO: understand the role of row-centering for getting circular patterns
## here, with completely random it doesn't really do it
##X <- X - apply(X, 1, mean) # ROW-CENTERING

X_centered <- scale(X, center = TRUE, scale = TRUE)  # Centered samples (genes as rows)


pca <- prcomp(X_centered, center = FALSE)  # PCA
svd_out <- svd(X_centered) # SVD


# Scores: projections of samples onto PCs
head(pca$x, 2)
# U * Sigma = same as PCA scores
head(svd_out$u %*% diag(svd_out$d), 2)

# Loadings: which genes define the PCs
head(pca$rotation, 2)
# V = right singular vectors = directions in sample space = PCA loadings
head(svd_out$v, 2)

# PCA scores ≈ U * Sigma
all.equal(c(pca$x), c(svd_out$u %*% diag(svd_out$d)), tolerance = 1e-15)

# PCA loadings ≈ V
all.equal(c(pca$rotation), c(svd_out$v), tolerance = 1e-15)

par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 2))
plot(pca$x[,1:2], main = "PCA: Sample Scores")
plot(svd_out$u[,1:2] %*% diag(svd_out$d[1:2]),
     main = expression("SVD:"~U * Sigma),
     xlab = expression(U[1] * Sigma[1]),
     ylab = expression(U[2] * Sigma[2]))

plot(pca$rotation[,1:2], main = "PCA: Loadings")
arrows(x0 = 0, y0 = 0, x1 = pca$rotation[,1], y1 = pca$rotation[,2],
       length = .1)
plot(svd_out$v[,1:2],
     main = expression(vectors),
     xlab = expression(V[1]),
     ylab = expression(V[2]))
arrows(x0 = 0, y0 = 0, x1 = svd_out$v[,1], y1 = svd_out$v[,2], length = .1)
```

# Circular Patterns in Rotated Data and Eigenvectors

TODO: find exact conditions to get circular data. Note: in single cell
data with just two cohort means we get 180degree eigenvectors even if
using co-expressed cohorts, with 3 60degree. Test this with sine wave
simulation from cohorts_simulate.R.


# R's `prcomp` explained

* Show equivalence for prcomp, eigen and Rspectra::eigen.
* stats::prcomp uses base::svd uses La.svd, based on LAPACK routines
 ‘DGESDD’ and ‘ZGESDD’.

In `prcomp`:

```
center = TRUE	Subtract mean from each column (variable)
scale. = TRUE	Subtract mean and divide by standard deviation (standardize)
```

where standardization is useful for heterogeneous data.



```{r plot, echo=TRUE, include=TRUE, fig.width=9, fig.height=3}
par(mfcol = c(1, 3))
script_path <- "/home/raim/programs/rcycle/inst/scripts/li06_plot.R"
#source(script_path)
```

# References
