---
title: "Piecewise Linear Segmentation by Dynamic Programming"
author: "Rainer Machn&eacute;"
date: "`r format(Sys.time(), '%d %m %Y')`"
output:
    bookdown::html_document2:
        base_format: rmarkdown::html_vignette
        toc: true
        toc_depth: 2
        fig_caption: true
bibliography: /home/raim/ref/tata.bib
vignette: >
  %\VignetteIndexEntry{Piecewise Linear Segmentation by Dynamic Programming}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<!-- compile with
rmarkdown::render("vignettes/rcycle.Rmd", output_format = "html_document")
-->

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, global.par=TRUE,
                      fig.path = ".", fig.pos = 'h',
                      fig.height = 2.7, fig.width = 4, fig.align = "center")
knitr::opts_knit$set(global.par = TRUE)
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25, cex.main=.75)
require(rcycle)
require(segmenTools)
source("~/programs/rcycle/R/phases.R")
source("~/programs/rcycle/R/plot.R")
```





$$
\newcommand{\Ell}{\mathcal{L}}
\newcommand{\jump}{\mathcal{J}}
\newcommand{\Var}{\mathrm{Var}}
\newcommand{\Cov}{\mathrm{Cov}}
\newcommand{\lmax}{\ell_\text{max}}
\newcommand{\lmin}{\ell_\text{min}}
\newcommand{\def}{\stackrel{\mathrm{def}}{=}}
$$

TODO: FILL THIS WITH teaching/pca/pca.md, teaching/pseudophase/cohorts14.md and clusterTranscriptomes_understandPCA.R

# Singular Value Decomposition and Principal Component Analysis

```{r load, echo=FALSE, include=TRUE}
## run pca in various versions, using data by Li and Klevecz (2006)
script_path <- system.file("scripts", "li06_pca.R", package = "yourpackage")
script_path <- "/home/raim/programs/rcycle/inst/scripts/li06_pca.R"
if ( !exists("cphase") )
source(script_path)
```

```{r}
### GENE-CENTERED
## VARIABLES/COLUMNS are samples/cells, OBSERVATIONS/ROWS are genes
## row-normalized and UNTRANSPOSED
gdat <- pdat - apply(pdat, 1, mean)
gphase <- prcomp(gdat, scale.=TRUE)

### SAMPLE-CENTERED with genes as columns!
## VARIABLES/COLUMNS are genes, OBSERVATIONS/ROWS are samples/cells

## row-norm over cells
sdat <- t(pdat) # transpose BEFORE row-norm 
sdat <- sdat - apply(sdat, 1, mean)  # row-norm over cells!
sphase <- prcomp(sdat, scale.=TRUE)

## row-norm over genes: as in @Schwabe2000?
cdat <- pdat - apply(pdat, 1, mean)  # row-norm over genes
cdat <- t(cdat) # transpose AFTER row-norm 
cphase <- prcomp(cdat, scale.=FALSE)  # NOTE: -> colnorm!

## pseudophases for all approaches
gtheta <- atan2(gphase$rotation[,2], gphase$rotation[,1])
stheta <- atan2(sphase$x[,2], sphase$x[,1])
ctheta <- atan2(cphase$x[,2], cphase$x[,1])
```

# The `biplot`

* @Gabriel_1971
* (https://stats.stackexchange.com/questions/66926/what-are-the-four-axes-on-pca-biplot)
* (https://stats.stackexchange.com/questions/141085/positioning-the-arrows-on-a-pca-biplot)


```{r, echo=TRUE, fig.width=7, fig.height=3.5}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 2))

## TODO: test biplot-like scaling more exactly!
## biplot w/o scaling
plotPC(gphase, arrows=FALSE, scores=TRUE, time.line=TRUE, eigen.axis=TRUE,
       arcsinh=FALSE, col=2, pch=19, cex=1)
par(new=TRUE)
biplot(gphase, scale=0, xlabs = rep("x", nrow(gphase$x))) 
figlabel('unscaled', pos='bottomleft', font=2)

## biplot with scaling
plotPC(gphase, arrows=FALSE, scores=TRUE, time.line=TRUE, eigen.axis=TRUE,
       scale=1, arcsinh=FALSE, col=2, pch=19, cex=1)
par(new=TRUE)
biplot(gphase, scale=1, xlabs = rep("x", nrow(gphase$x)))
figlabel('true biplot', pos='bottomleft', font=2)
```

# Gene vs. Sample PCA


```{r, echo=TRUE, fig.width=9, fig.height=3}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))

## sample PCA: with samples as variables/columns - most unconventional
plotPC(gphase, arrows=FALSE, scores=TRUE, time.line=TRUE, eigen.axis=TRUE,
       col=2, pch=19, cex=1) # eigenvectors/rotation matrix are samples
figlabel('sample PCA', pos='bottomleft', font=2)

## gene PCA: with genes as variables/columns
## but with row-norm over cells after transpose
plotPC(sphase, arrows=FALSE, ccol=2, cpch=19, eigen.axis=TRUE)
lines(sphase$x[,1], sphase$x[,2], col=2) # rotated data are samples
figlabel('gene PCA, cell-norm', pos='bottomleft', font=2)

## gene PCA: with genes as variables/columns - conventional
## with row-norm over genes prior to transpose
plotPC(cphase, arrows=FALSE, ccol=2, cpch=19, eigen.axis=TRUE)
lines(cphase$x[,1], cphase$x[,2], col=2) # rotated data are samples
figlabel('gene PCA, gene-norm', pos='bottomleft', font=2)
```

```{r, echo=TRUE}
## COMPARE PSEUDOPHASES
## pseudophases are very similar between approaches
gtheta <- atan2(gphase$rotation[,2], gphase$rotation[,1])
stheta <- atan2(sphase$x[,2], sphase$x[,1])
ctheta <- atan2(cphase$x[,2], cphase$x[,1])
```

```{r, echo=FALSE, fig.width=9, fig.height=3}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))
plot(time, -gtheta, xlab='time/h', ylab=expression(-theta~'sample PCA'))
plot(time, ctheta, xlab='time/h', ylab=expression(theta~'gene PCA, cell-norm'))
plot(time, stheta, xlab='time/h', ylab=expression(theta~'gene PCA, gene-norm'))
```

```{r, echo=FALSE, fig.width=9, fig.height=3}
par(mai=c(.5,.5,.5,.5),mgp=c(1.3,.3,0),tcl=-.25)
par(mfcol = c(1, 3))
plot(-gtheta, stheta, xlab=expression(-theta~'sample PCA'), ylab=expression(theta~'gene PCA, cell-norm'))
plot(-gtheta, ctheta, xlab=expression(-theta~'sample PCA'), ylab=expression(theta~'gene PCA, gene-norm'))
plot(stheta, ctheta, xlab=expression(theta~'gene PCA, cell-norm'), ylab=expression(theta~'gene PCA, gene-norm'))
```

TODO:

* same on cohort level!
* single cell data: are we using with or w/o rownorm?
* take log of counts, esp. for single cell where arcsinh scaling
  is required.


# Time Series Reconstruction: `pseudophase`.

eigengenes @Alter2000,

applied for single cell analysis: @Schwabe2020, @Macosko2015


# R's `prcomp` explained

Show equivalence for prcomp, eigen and Rspectra::eigen.

```{r plot, echo=TRUE, include=TRUE, fig.width=9, fig.height=3}
par(mfcol = c(1, 3))
script_path <- "/home/raim/programs/rcycle/inst/scripts/li06_plot.R"
source(script_path)
```

# References
